FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/FreeRTOS_IP.c

```diff
@@ -2198,10 +2199,20 @@ size_t uxDataLengthBytes = uxByteCount;
 	xSource.u8ptr = ipPOINTER_CAST( uint8_t *, pucNextData );
 	/* coverity[misra_c_2012_rule_11_4_violation] */
 	/* The object pointer expression "pucNextData" of type "uint8_t const *" is cast to an integer type "unsigned int". */
-	ulAlignBits = ( ( ( uint32_t ) pucNextData ) & 0x03U ); /*lint !e9078 !e923*/	/* gives 0, 1, 2, or 3 */
+	uxAlignBits = ( ( intptr_t ) pucNextData ) & 0x03U; /*lint !e9078 !e923*/	/* gives 0, 1, 2, or 3 */
+
+	/*
+	 * If pucNextData is non-aligned then the checksum is starting at an
+	 * odd position and we need to make sure the usSum value now in xSum is
+	 * as if it had been "aligned" in the same way.
+	 */
+	if( ( uxAlignBits & 1UL) != 0U )
+	{
+		xSum.u32 = ( ( xSum.u32 & 0xffU ) << 8 ) | ( ( xSum.u32 & 0xff00U ) >> 8 );
```

Comment:
Why is this using a 32-bit data type given that the checksum is only 16 bits?

Reply:
I had the same question!
My research suggests that this is a common algorithm for calculating the checksum that gives some performance benefits on machines with word-sizes greater than 16 bits.

Part (C) "Parallel Summation" at https://tools.ietf.org/html/rfc1071#page-3 has some more detail.

The overall algorithm implementation is pre-existing work, it just seems to have an alignment bug on 8-bit platforms: if the additions are done in this parallel way then it's important to make sure that the correct parts of the packet end up in the correct part of the summation: as long as it's consistent it doesn't matter how they're arranged (Part (B) at the above link), but on 8-bit platforms the arrangement was inconsistent.

Reply:
> Why is this using a 32-bit data type given that the checksum is only 16 bits?

Yes, pure optimisation.  The algorithm tries to do as many bytes in a single operation as possible.
So it will look for a 16-bit alignment, and then to a 32-bit alignment.  Bytes will then be added four at s time ("Parallel Summation").

The tricky thing is: the alignment can only be calculated when a pointer is cast into a number. MISRA doesn't like that.  What I did is cast it to a `uin32_t`, @andyjpb proposes to cast it to a intptr_t in stead, which has the same size as a void *, while bit-tests are allowed. So I vote for this change.